docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+coupling between images. Likewise, if an image requires configuration in order
+to start and that configuration is environment dependent, the image must be
+altered to add the necessary templating or retrieval.
+
+This proposal introduces the concept of an **init container**, one or more
+containers started in sequence before the pod's normal containers are started.
+These init containers may share volumes, perform network operations, and perform
+computation prior to the start of the remaining containers. They may also, by
+virtue of their sequencing, block or delay the startup of application containers
+until some precondition is met. In this document we refer to the existing pod
+containers as **app containers**.
+
+## Design Points
+
+* Init containers should be able to:
+  * Perform initialization of shared volumes
+  * Delay the startup of application containers until preconditions are met
+  * Download binaries that will be used in app containers as execution targets
+  * Inject configuration or extension capability to generic images at startup
+  * Perform complex templating of information avaliable in the local environment
+  * Register the pod with other components of the system
+* Reduce coupling between application images and reduce the need to customize
+  those images for Kubernetes generally or specific roles
+* Reduce the footprint of individual images by specializing which containers
+  perform which tasks (install git into init container, use filesystem contents
+  in web container)
+* The order init containers start should be predictable and allow users to easily
+  reason about the startup of a container
+* Complex ordering and failure is out of scope - all complex workflows can if
+  necessary be implemented inside of a single init container, and this proposal
+  aims to enable that ordering without adding undue complexity to the system
+* Both run-once and run-forever pods should be able to use init containers
+* As much as possible, an init container should behave like an app container
+  to reduce complexity for end users, for clients, and for divergent use cases.
+  An init container is a container with the minimum alterations to accomplish
+  its goal.
+
+## Alternatives
+
+* Any mechanism that runs user code on a node before regular pod containers
+  should itself be a container and modeled as such - we explicitly reject
+  creating new mechanisms for running user processes.
+* The container pre-start hook (not yet implemented) requires execution within
+  the container's image and so cannot adapt existing images. It also cannot
+  block startup of containers
+* Running a "pre-pod" would defeat the purpose of the pod being an atomic
+  unit of scheduling.
+
+
+## Design
+
+Each pod may have 0..N init containers defined along with the existing
+1..N app containers.
+
+On startup of the pod, after the network and volumes are initialized, the
+init containers are started in order. Each container must exit successfully
```

Comment:
I think we should think about what the ideal end state is. It's not clear to me that the implementation of this model is that much simpler than startup dependencies, and it might add an unnecessary maintenance burden. On the other hand, we may not want to accept the full burden of responsibility that comes with the graph model (e.g. loop detection? Or do we want to allow a turing-complete pod lifecycle :P )